### PR TITLE
[JENKINS-51594] Test behavior of snapshots vs. RCs, JEP-229

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-51594](https://issues.jenkins-ci.org/browse/JENKINS-51594)

See https://github.com/jenkinsci/jenkins/pull/153 where @stephenc was apparently trying to act like Maven by default.

### CODE REVIEW

File: src/main/java/hudson/util/VersionNumber.java [hunk 1]
```diff
@@ -48,10 +48,6 @@
  * <p>
  * 'SNAPSHOT' is also allowed as a component, and "N.SNAPSHOT" is interpreted as "N-1.*"
  *
- * <pre>
- * 2.0.* &lt; 2.0.1 &lt; 2.0.1-SNAPSHOT &lt; 2.0.0.99 &lt; 2.0.0 &lt; 2.0.ea &lt; 2.0
```
Comment (jglick) [inline src/main/java/hudson/util/VersionNumber.java]: This was just so far off that I did not see how to rescue it. 2.0.1 < 2.0??

**Comment by jglick** [comment]
CC @oleg-nenashev @batmat

File: src/main/java/hudson/util/VersionNumber.java [hunk 2]
```diff
@@ -180,7 +176,7 @@ public String toString() {
      * Represents a string in the version item list, usually a qualifier.
      */
     private static class StringItem implements Item {
-        private final static String[] QUALIFIERS = {"snapshot", "alpha", "beta", "milestone", "rc", "", "sp"};
+        private final static String[] QUALIFIERS = {"alpha", "beta", "milestone", "rc", "snapshot", "", "sp"};
```
Comment (stephenc) [inline src/main/java/hudson/util/VersionNumber.java]: Technically from Maven's PoV `-SNAPSHOT` happens before `-alpha-SNAPSHOT` which happens before the release.

IOW your version starts off as 1.1-SNAPSHOT, then goes to 1.1-alpha-1-SNAPSHOT, 1.1-alpha-1, 1.1-alpha-2-SNAPSHOT, 1.1-alpha-2, 1.1-beta-1-SNAPSHOT, 1.1-beta-1, 1.1-rc-1-SNAPSHOT, 1.1-rc-1, ..., 1.1, 1.1-sp-1-SNAPSHOT, 1.1-sp-1, 1.1-sp-2-SNAPSHOT

So if you want to stay consistent with Maven's version handling, this is not the change you want

Comment (jglick) [reply src/main/java/hudson/util/VersionNumber.java]: Hmm…I am wondering whether we should just be making this library a shim for `ComparableVersion`. It is not entirely clear to me why it differs at all.

Comment (stephenc) [reply src/main/java/hudson/util/VersionNumber.java]: some "magic" that KK had for `*` as a version component

Comment (jglick) [reply src/main/java/hudson/util/VersionNumber.java]: I saw a comment to that effect…but is this “feature” actually being used anywhere that you know of?

Comment (stephenc) [reply src/main/java/hudson/util/VersionNumber.java]: I know not of any usage of that...

Comment (jtnord) [reply src/main/java/hudson/util/VersionNumber.java]: Hmm milestones are usually pre-alpha  thus `milestone -> alpha -> beta -> rc -> ga` but yeah maven ¯\_(ツ)_/¯

**Comment by jglick** [comment]
> this is now conflicted

Fixed.

> Do you plan to finish it?

Well it was not “unfinished”, it was more that there was no clear consensus on what the behavior of this library should be or what the historical decisions were.

File: src/main/java/hudson/util/VersionNumber.java (hunk 2 above)
Comment (jtnord) [reply src/main/java/hudson/util/VersionNumber.java:179]: for reference https://maven.apache.org/ref/3.6.0/maven-artifact/apidocs/org/apache/maven/artifact/versioning/ComparableVersion.html

**Review by jtnord — COMMENTED** [review]
as per Stephenc. 

changing the order of rc and snapshot whilst makes sense for our incrementals JEP (222?) it does not match mavens ordering and I am not clear on what if any issues that may cause.

As Snapshots are normally clearly identified and can be filtered this should not cause too many issues (famous last words), but I am not clear on where any checking would actually take place between a snapshot and a non snapshot version.

**Comment by jglick** [comment]
> it does not match mavens ordering

Right: https://github.com/jenkinsci/lib-version-number/pull/6#discussion_r193770807

https://github.com/jenkinsci/jenkins/pull/3551 makes this less interesting.

File: src/test/java/hudson/util/VersionNumberTest.java [hunk 3]
```diff
@@ -88,6 +113,28 @@ public void testDigit() {
...
+        assertOrderAlsoInMaven("2.0.0", "2.0", "2.0.*", "2.0.ea", "2.0.0.99", "2.0.1-alpha-1-rc9999.abc123def456", "2.0.1-alpha-1", "2.0.1-rc9999.abc123def456", "2.0.1-SNAPSHOT", "2.0.1");
+    }
+
+    @Issue("JEP-229")
+    @Test
+    public void backportJep229() {
+        // Maven considers 99.1.abcd1234abcd to sort before 99.1234deadbeef so we cannot simply use 99. as the branch prefix.
+        // Nor can we use 99.1234deadbeef. as the prefix because Maven would compare 5 and 10 lexicographically.
+        // 100._. seems to work but is not intuitive.
+        // Using changelist.format=%d.v%s behaves better, apparently because then the hash is never treated like a number.
```
Comment (jglick) [inline src/test/java/hudson/util/VersionNumberTest.java:134]: https://github.com/jenkinsci/incrementals-tools/pull/19

**Review by timja — APPROVED** [review]

**Pull request merged**